app/controllers/actions/IdentifierAction.scala

```diff
@@ -137,19 +138,70 @@ class AuthenticatedIdentifierAction @Inject()(
       case Some(role) =>
         getLoggedInUser(externalId, role, enrolments) match {
-          case Some(loggedInUser) => block(IdentifierRequest(request, loggedInUser, eventReporting.pstr, eventReporting.schemeName, eventReporting.returnUrl, eventReporting.srn))
+          case Some(loggedInUser) =>
+            val userAuthorised = for {
+              psaId <- getPsaId(enrolments)
+              pspId <- getPspId(enrolments)
+            } yield {
+              for {
+                psaAuthorised <- schemeAuthorisedForPsa(psaId, eventReporting.srn)
+                pspAuthorised <- schemeAuthorisedForPsp(pspId, eventReporting.pstr)
+              } yield {
+                psaAuthorised || pspAuthorised
+              }
+            }
+
+            userAuthorised.map { authFtr =>
+              authFtr.flatMap {
+                case true => block(IdentifierRequest(request, loggedInUser, eventReporting.pstr, eventReporting.schemeName, eventReporting.returnUrl, eventReporting.srn))
+                case false => futureUnauthorisedPage
+              }
+            }.getOrElse(futureUnauthorisedPage)
+
           case _ => futureUnauthorisedPage
         }
     }
   }
 
-  def actionForOneEnrolment[A](administratorOrPractitioner: AdministratorOrPractitioner,
+  private def actionForOneEnrolment[A](administratorOrPractitioner: AdministratorOrPractitioner,
                                eventReporting: EventReporting,
                                externalId: String, enrolments: Enrolments, request: Request[A],
-                               block: IdentifierRequest[A] => Future[Result]): Future[Result] = {
+                               block: IdentifierRequest[A] => Future[Result])(implicit headerCarrier: HeaderCarrier): Future[Result] = {
+
     getLoggedInUser(externalId, administratorOrPractitioner, enrolments) match {
-      case Some(loggedInUser) => block(IdentifierRequest(request, loggedInUser, eventReporting.pstr, eventReporting.schemeName, eventReporting.returnUrl, eventReporting.srn))
+      case Some(loggedInUser) =>
+        def proceed = block(IdentifierRequest(request, loggedInUser, eventReporting.pstr, eventReporting.schemeName, eventReporting.returnUrl, eventReporting.srn))
+        administratorOrPractitioner match {
+          case AdministratorOrPractitioner.Administrator =>
+            schemeAuthorisedForPsa(loggedInUser.psaIdOrPspId, eventReporting.srn).flatMap {
+              case true => proceed
+              case false => futureUnauthorisedPage
+            }
+          case AdministratorOrPractitioner.Practitioner => schemeAuthorisedForPsp(loggedInUser.psaIdOrPspId, eventReporting.pstr).flatMap {
+            case true => proceed
+            case false => futureUnauthorisedPage
+          }
+        }
+
       case _ => futureUnauthorisedPage
     }
   }
+
+  private def schemeAuthorisedForPsa(psaId: String, srn: String)(implicit headerCarrier: HeaderCarrier): Future[Boolean] = {
```

Comment:
It's pstr for PSP because that's how it was implemented for ER for whatever reason and I wanted to make sure that the changes are make are minimal so it's easier for us to review.
Practically speaking, there should not be a difference in us using PSTR or SRN.